app/models/file_depot_swift.rb

```diff
@@ -0,0 +1,73 @@
+class FileDepotSwift < FileDepot
+  attr_accessor :swift
+
+  def self.uri_prefix
+    "swift"
+  end
+
+  def self.validate_settings(settings)
+    new(:uri => settings[:uri]).verify_credentials(nil, settings.slice(:username, :password))
+  end
+
+  def connect(options = {})
+    uri  = options[:uri]
+    host = URI(uri).host
+    openstack_handle(options).connect(options)
+  rescue Excon::Errors::Unauthorized => err
+    logger.error("Access to Swift host #{host} failed due to a bad username or password. #{err}")
+    nil
+  rescue => err
+    logger.error("Error connecting to Swift host #{host}. #{err}")
+    msg = "Error connecting to Swift host #{host}. #{err}"
+    raise err, msg, err.backtrace
+  end
+
+  def openstack_handle(options = {})
+    require 'manageiq/providers/openstack/legacy/openstack_handle'
+    @openstack_handle ||= begin
+      username = options[:username] || authentication_userid(options[:auth_type])
+      password = options[:password] || authentication_password(options[:auth_type])
+      uri      = options[:uri]
+      address  = URI(uri).host
+      port     = URI(uri).port
+
+      extra_options = {
+        :ssl_ca_file    => ::Settings.ssl.ssl_ca_file,
+        :ssl_ca_path    => ::Settings.ssl.ssl_ca_path,
+        :ssl_cert_store => OpenSSL::X509::Store.new
+      }
+      extra_options[:domain_id]         = v3_domain_ident
+      extra_options[:service]           = "Compute"
+      extra_options[:omit_default_port] = ::Settings.ems.ems_openstack.excon.omit_default_port
+      extra_options[:read_timeout]      = ::Settings.ems.ems_openstack.excon.read_timeout
+      begin
+        OpenstackHandle::Handle.new(username, password, address, port, keystone_api_version, security_protocol, extra_options)
+      rescue => err
+        logger.error("Error connecting to Swift host #{address}. #{err}")
+        msg = "Error connecting to Swift host #{address}. #{err}"
+        raise err, msg, err.backtrace
+      end
+    end
+  end
+
+  def verify_credentials(auth_type = nil, options = {})
+    auth_type ||= 'default'
+
+    options[:auth_type] = auth_type
+    connect(options.merge(:auth_type => auth_type))
+  rescue => err
+    logger.error("Error connecting to Swift host #{host}. #{err}")
+    msg = "Error connecting to Swift host #{host}. #{err}"
+    raise err, msg, err.backtrace
+  end
+
+  def merged_uri(uri, api_port)
+    port      = api_port.presence || 5000
+    uri       = URI.parse("#{URI(uri).scheme}://#{URI(uri).host}:#{port}#{URI(uri).path}")
+    uri.query = [uri.query, "region=#{openstack_region}"].compact.join('&') if openstack_region.present?
+    uri.query = [uri.query, "api_version=#{keystone_api_version}"].compact.join('&') if keystone_api_version.present?
+    uri.query = [uri.query, "domain_id=#{v3_domain_ident}"].compact.join('&') if v3_domain_ident.present?
+    uri.query = [uri.query, "security_protocol=#{security_protocol}"].compact.join('&') if security_protocol.present?
```

Comment:
I think we could simplify this a bit by building a list of query elements outside of the `uri.query` string object. Like this:

```ruby
query_elements = []
query_elements << "region=#{openstack_region}"              if openstack_region.present?
query_elements << "api_version=#{keystone_api_version}"     if keystone_api_version.present?
query_elements << "domain_id=#{v3_domain_ident}"]           if v3_domain_ident.present?
query_elements << "security_protocol=#{security_protocol}"] if security_protocol.present?
uri.query = query_elements.join('&').presence
```

Also we can drop the `.present?` calls if these will only ever be nil or usable.

Reply:
If its nil then we'll end up with query strings like "region=&api_version=&domain_id=", won't we?

Reply:
Nope, I'm pretty sure my version works as I'm only adding the `key=value` if the value is present, then joining only when the array is fully assembled.

Also, I'm not sure what "it" you're referencing here, but we should probably have a test for this method either way. I would suggest writing some tests that cover `nil` values and then we can refactor with confidence.

Reply:
@carbonin I'm referring to your question about dropping the ".present?" calls if these will be nil or usable.  if they're nil, and there is no .present? call, then the wrong thing will happen.  Your code works correctly as written.

Reply:
> if they're nil, and there is no .present? call, then the wrong thing will happen. 

No, `nil` is falsey.

This won't print anything:

```ruby
foo = nil
puts "this won't print" if foo
```

What `present?` and `presence` do is move empty strings (and also other empty objects like arrays, hashes, etc) to also be `false` or `nil` respectively. So when I saw you using `present?` I assumed that these could possibly be empty strings, but that seemed strange, which is why I asked.